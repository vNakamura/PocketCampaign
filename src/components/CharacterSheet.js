import React, { Component } from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';

import { charChange } from '../actions/chat';
import LabeledInput from './Inputs/LabeledInput';
import Scrollable from './Scrollable';

const Box = styled.div`
  display: flex;
  flex-direction: column;
  max-height: 40vh;
`;

const Row = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
`;

export class CharacterSheet extends Component {
  props: {
    char: { [string]: string },
    dispatch: Function,
  }

  handleInputChange = (field: string, value: string): void => {
    this.props.dispatch(charChange({
      ...this.props.char,
      [field]: value,
    }));
  }
  render() {
    const { char } = this.props;
    return (
      <Box>
        <Scrollable>
          <Row>
            <LabeledInput
              label="Name"
              onChange={this.handleInputChange}
              inputProps={{
                size: 25,
                value: char.name || '',
              }}
              flex={4}
            />
            <LabeledInput
              label="Class & Level"
              onChange={this.handleInputChange}
              inputProps={{
                size: 10,
                value: char.classLevel || '',
              }}
            />
            <LabeledInput
              label="Race"
              onChange={this.handleInputChange}
              inputProps={{
                size: 10,
                value: char.race || '',
              }}
            />
            <LabeledInput
              label="Background"
              onChange={this.handleInputChange}
              inputProps={{
                size: 10,
                value: char.background || '',
              }}
            />
            <LabeledInput
              label="Alignment"
              onChange={this.handleInputChange}
              inputProps={{
                size: 10,
                value: char.alignment || '',
              }}
            />
          </Row>
          <Row>
            <LabeledInput
              label="Strength"
              onChange={this.handleInputChange}
              inputProps={{
                size: 10,
                value: char.strength || '',
              }}
            />
            <LabeledInput
              label="Dextery"
              onChange={this.handleInputChange}
              inputProps={{
                size: 10,
                value: char.dextery || '',
              }}
            />
            <LabeledInput
              label="Constitution"
              onChange={this.handleInputChange}
              inputProps={{
                size: 10,
                value: char.constitution || '',
              }}
            />
            <LabeledInput
              label="Intelligence"
              onChange={this.handleInputChange}
              inputProps={{
                size: 10,
                value: char.intelligence || '',
              }}
            />
            <LabeledInput
              label="Wisdom"
              onChange={this.handleInputChange}
              inputProps={{
                size: 10,
                value: char.wisdom || '',
              }}
            />
            <LabeledInput
              label="Charisma"
              onChange={this.handleInputChange}
              inputProps={{
                size: 10,
                value: char.charisma || '',
              }}
            />
          </Row>
        </Scrollable>
      </Box>
    );
  }
}

const mapStateToProps = state => ({
  char: state.chars.demoChar,
});
export default connect(mapStateToProps)(CharacterSheet);
