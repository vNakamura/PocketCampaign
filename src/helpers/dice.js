// @flow

import React from 'react';
import droll from 'droll';
import seedrandom from 'seedrandom';
import replace from 'string-replace-to-array';

import NotationLink from '../components/Chat/NotationLink';

export const notationLimits = {
  minNumDice: 1,
  maxNumDice: 10,
  minNumSides: 2,
  maxNumSides: 100,
  minModifier: -1000,
  maxModifier: 1000,
};
export const diceNotationRegex = /\b([1-9]\d*)?d([1-9]\d*)([+-]\d+)?\b/gi;

type ResultError = {
  ok: false,
  message: string,
};
type ValidateResult = ResultError | { ok: true };
const validateRoll = (notation:string): ValidateResult => {
  const p = droll.parse(notation);
  if (!p) return { ok: false, message: 'Invalid dice notation' };
  if (p.numDice > notationLimits.maxNumDice) return { ok: false, message: 'Invalid dice notation. Too many dice!' };
  if (p.numSides < notationLimits.minNumDice) return { ok: false, message: 'Invalid dice notation. You need at least one die!' };
  if (p.numSides > notationLimits.maxNumSides) return { ok: false, message: 'Invalid dice notation. Too many sides!' };
  if (p.numSides < notationLimits.minNumSides) return { ok: false, message: 'Invalid dice notation. Too few sides!' };
  if (p.modifier > notationLimits.maxModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  if (p.modifier < notationLimits.minModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  return { ok: true };
};

export type ParseResult = ResultError | {
  ok: true,
  parsed: {
    numDice: number,
    numSides: number,
    modifier: number,
    minResult: number,
    maxResult: number,
    avgResult: number,
  }
};
export const parseRoll = (notation: string): ParseResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;
  const p = droll.parse(notation);
  return { ok: true, parsed: p };
};

export type RollResult = ResultError | {
  ok: true,
  rolls: number[],
  modifier: number,
  total: number,
};
export const roll = (notation: string, seed?: string): RollResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;

  if (seed !== undefined) seedrandom(seed, { global: true });
  const result = droll.roll(notation);
  return {
    ok: true,
    rolls: result.rolls,
    modifier: result.modifier,
    total: result.total,
  };
};

export const addLinksToText = (text: string) => {
  let i = 0;
  const result = replace(
    text,
    diceNotationRegex,
    (notation, numDice, numSides) => {
      i += 1;
      const v: ValidateResult = validateRoll(notation);
      if (v.ok && numSides > 1) return <NotationLink key={i} notation={notation} />;
      return notation;
    },
  );
  return result;
};
