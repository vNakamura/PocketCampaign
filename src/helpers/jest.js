import React from 'react';
import { ThemeProvider } from 'styled-components';
/* eslint-disable import/no-extraneous-dependencies */
import renderer from 'react-test-renderer';
import 'jest-styled-components';
/* eslint-enable import/no-extraneous-dependencies */
import theme from '../theme';

export const applyTheme = content => (
  <ThemeProvider theme={theme}>
    {content}
  </ThemeProvider>
);

export const renderWithTheme = content => renderer.create(
  applyTheme(content),
).toJSON();

export default {
  applyTheme,
  renderWithTheme,
};
