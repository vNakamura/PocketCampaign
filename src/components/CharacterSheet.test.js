import React from 'react';

import { CharacterSheet } from './CharacterSheet';
import { renderWithTheme } from '../helpers/jest';

describe('components: Inputs/CharacterSheet', () => {
  it('should render correctly', () => {
    const demo = renderWithTheme(<CharacterSheet char={{}} />);
    expect(demo).toMatchSnapshot();
  });
});
