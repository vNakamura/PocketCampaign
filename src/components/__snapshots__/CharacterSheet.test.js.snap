// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components: Inputs/CharacterSheet should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-height: 40vh;
}

.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow-y: auto;
  -webkit-overflow-scrolling: touch;
  display: block;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 4;
  -ms-flex: 4;
  flex: 4;
  margin: .5rem;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: .5rem;
}

.c4 {
  font-family: "Open Sans";
  font-size: 1rem;
  padding: .4em;
  border: none;
  border-radius: 0;
  text-align: left;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        Name
        <input
          className="c4"
          onChange={[Function]}
          size={25}
          value=""
        />
      </div>
      <div
        className="c5"
      >
        Class & Level
        <input
          className="c4"
          onChange={[Function]}
          size={10}
          value=""
        />
      </div>
      <div
        className="c5"
      >
        Race
        <input
          className="c4"
          onChange={[Function]}
          size={10}
          value=""
        />
      </div>
      <div
        className="c5"
      >
        Background
        <input
          className="c4"
          onChange={[Function]}
          size={10}
          value=""
        />
      </div>
      <div
        className="c5"
      >
        Alignment
        <input
          className="c4"
          onChange={[Function]}
          size={10}
          value=""
        />
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c5"
      >
        Strength
        <input
          className="c4"
          onChange={[Function]}
          size={10}
          value=""
        />
      </div>
      <div
        className="c5"
      >
        Dextery
        <input
          className="c4"
          onChange={[Function]}
          size={10}
          value=""
        />
      </div>
      <div
        className="c5"
      >
        Constitution
        <input
          className="c4"
          onChange={[Function]}
          size={10}
          value=""
        />
      </div>
      <div
        className="c5"
      >
        Intelligence
        <input
          className="c4"
          onChange={[Function]}
          size={10}
          value=""
        />
      </div>
      <div
        className="c5"
      >
        Wisdom
        <input
          className="c4"
          onChange={[Function]}
          size={10}
          value=""
        />
      </div>
      <div
        className="c5"
      >
        Charisma
        <input
          className="c4"
          onChange={[Function]}
          size={10}
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
